File: Add Code Here/cpp/linked list reverse.cpp
#include<iostream>
using namespace std;
class node{
public:
	int data;
	node *next;
	
	node(int d){
		data=d;
		next=NULL;
	}
};


void insert_at_head(node *&head,int d){
	if(head==NULL){
		head=new node(d);
		return;
	}
	node *n=new node(d);
	n->next=head;
	head=n;
	
}
void print(node *head){
	while(head!=NULL){
		cout<<head->data<<"--->";
		head=head->next;
	}
}
void reverse(node *&head){
	node *current=head;
	node *prev=NULL;
	node *n;
	while(current!=NULL){
		n=current->next;    //save the next node
		current->next=prev;  //make the current node point to previous
		prev=current;  //update previous 1 step forward
		current=n;    //update current 1 step forward
	}
	head=prev;
}
int main(){
	node *head=NULL;
	insert_at_head(head,3);
	insert_at_head(head,2);
	insert_at_head(head,1);
	insert_at_head(head,0);
	
	print(head);
	
	reverse(head);
	
	cout<<endl;
	
	print(head);
	
}

File: Add Code Here/JAVA/LinkedListReverse.java
package LinkedList;
import java.util.*;

public class LinkedListReverse
{	
	Node head;
	static class Node
	{
		Node next;
		int data;
		Node(int d)
		{	data=d;	}		
	}
	
	public static LinkedListReverse insert(LinkedListReverse list,int data)
	{
		Node newNode=new Node(data);
		newNode.next=null;
		if(list.head==null)
		{	list.head=newNode;	}
		else
		{	Node last=list.head;
			while(last.next!=null)
			{	last=last.next;		}
			last.next=newNode;
		}
		return list;
	}
	
	public static void display(Node node)
	{ 
		while (node != null) 
		{ 	System.out.print(node.data+"  "); 
			node = node.next; 	} 
	}
	
	public static Node reverse(Node head)
	{
		if(head==null || head.next==null)
		{	return head;	}				
		Node newHeadNode=reverse(head.next);
		head.next.next=head;
		head.next=null;
		
		return newHeadNode;
	}
	
	public static void main(String[] args)
	{
		Scanner sc=new Scanner(System.in);
		System.out.print("Enter no of nodes:");
		int n=sc.nextInt();
		LinkedListReverse list=new LinkedListReverse();
		System.out.print("Enter elements:");
		for(int i=0;i<n;i++)
		{
			int x=sc.nextInt();
			insert(list,x);
		}
		System.out.print("LinkedList: ");
		display(list.head);
		Node list1=reverse(list.head);
		System.out.print("\nReverse LinkedList: ");
		display(list1);
	}
}
